File: packages/ui/src/editor/view/native/android/src/main/java/com/kryton/editor/KrytonEditorManager.kt
// packages/ui/src/editor/view/native/android/src/main/java/com/kryton/editor/KrytonEditorManager.kt
package com.kryton.editor

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class KrytonEditorManager : SimpleViewManager<KrytonEditor>() {
    override fun getName() = "KrytonEditor"

    override fun createViewInstance(reactContext: ThemedReactContext): KrytonEditor =
        KrytonEditor(reactContext)

    @ReactProp(name = "text")
    fun setText(view: KrytonEditor, text: String?) {
        view.applyText(text ?: "")
    }

    @ReactProp(name = "decorations")
    fun setDecorations(view: KrytonEditor, decorations: ReadableArray?) {
        val list = mutableListOf<DecorationSpec>()
        decorations ?: run { view.applyDecorations(emptyList()); return }
        for (i in 0 until decorations.size()) {
            val m: ReadableMap = decorations.getMap(i) ?: continue
            val attrsMap = m.getMap("attrs")
            val attrs = if (attrsMap != null) {
                val it = attrsMap.keySetIterator()
                val map = mutableMapOf<String, String>()
                while (it.hasNextKey()) {
                    val k = it.nextKey()
                    map[k] = attrsMap.getString(k) ?: continue
                }
                map
            } else null
            list.add(DecorationSpec(
                from = m.getInt("from"),
                to = m.getInt("to"),
                kind = m.getString("kind") ?: continue,
                attrs = attrs,
            ))
        }
        view.applyDecorations(list)
    }

    @ReactProp(name = "selection")
    fun setSelection(view: KrytonEditor, sel: ReadableMap?) {
        sel ?: return
        view.applySelection(sel.getInt("anchor"), sel.getInt("head"))
    }

    override fun getExportedCustomBubblingEventTypeConstants(): MutableMap<String, Any> {
        val phasedRegistration = mapOf("bubbled" to "onChangeText", "captured" to "onChangeTextCapture")
        return mutableMapOf(
            "topChange" to mapOf("phasedRegistrationNames" to phasedRegistration),
            "topSelectionChange" to mapOf("phasedRegistrationNames" to mapOf("bubbled" to "onChangeSelection", "captured" to "onChangeSelectionCapture")),
            "topWikilinkPress" to mapOf("phasedRegistrationNames" to mapOf("bubbled" to "onWikilinkPress", "captured" to "onWikilinkPressCapture")),
        )
    }
}

File: packages/ui/src/editor/view/native/ios/KrytonEditorManager.swift
// packages/ui/src/editor/view/native/ios/KrytonEditorManager.swift
import React

@objc(KrytonEditorManager)
public final class KrytonEditorManager: RCTViewManager {
    public override func view() -> UIView! {
        return KrytonEditor(frame: .zero, textContainer: nil)
    }

    public override static func requiresMainQueueSetup() -> Bool { return true }

    @objc public func updateText(_ reactTag: NSNumber, text: NSString) {
        bridge?.uiManager.addUIBlock { (_, viewRegistry) in
            if let view = viewRegistry?[reactTag] as? KrytonEditor {
                view.setText(text as String)
            }
        }
    }

    @objc public func updateDecorations(_ reactTag: NSNumber, decorations: NSArray) {
        bridge?.uiManager.addUIBlock { (_, viewRegistry) in
            if let view = viewRegistry?[reactTag] as? KrytonEditor {
                view.setDecorations(decorations)
            }
        }
    }

    @objc public func updateSelection(_ reactTag: NSNumber, anchor: NSNumber, head: NSNumber) {
        bridge?.uiManager.addUIBlock { (_, viewRegistry) in
            if let view = viewRegistry?[reactTag] as? KrytonEditor {
                view.setSelection(anchor.intValue, head.intValue)
            }
        }
    }
}
